---
title: "Proportion Test"
author: "Yaamini Venkataraman"
date: "1/15/2019"
output: html_document
---

I will use a proportion test to compare the proportion of genome feature overlaps between differentially methylated loci (DML), differentially methylated regions (DMR), and the gene background.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Obtain session information

```{r}
sessionInfo()
```

# Import data

```{r}
overlapData <- read.csv("2019-01-15-Overlap-Proportions.csv", header = TRUE)
rownames(overlapData) <- overlapData$genomicFeature #Set genomic feature indication and rownames
overlapData <- overlapData[,-1] #Remove genomic feature indication column
head(overlapData) #Confirm import
```

# Conduct chi-squared tests of homogeneity

The null hypothesis is that loci distributions in the genome are the same between different CpG categories.

## Total CpGs vs. methylated CpGs

```{r}
totalVersusMethylated <- overlapData[, 1:2] #Subset data and transform for statistical testing
totalVersusMethylated <- totalVersusMethylated[-6, ] #Remove totalLines row
head(totalVersusMethylated) #Confirm changes
```

```{r}
totalVersusMethylatedTest <- chisq.test(totalVersusMethylated) #Conduct a chi-squared test
totalVersusMethylatedTest #The distribution of methylated CpGs is significantly different from all CpGs in the C. virginica genome.
```

## Methylated CpGs vs. DML

```{r}
methylatedVersusDML <- overlapData[,2:3] #Subset data and transform for statistical testing
methylatedVersusDML <- methylatedVersusDML[-6,] #Remove totalLines row
head(methylatedVersusDML) #Confirm changes
```

```{r}
methylatedVersusDMLTest <- chisq.test(methylatedVersusDML) #Conduct a chi-squared test
methylatedVersusDMLTest #The distribution of DML is significantly different from methylated CpGs.
```

# Conduct post-hoc test

# Create figures

## Reformat data

```{r}
head(overlapData)
```


```{r}
proportionData <- overlapData #Copy overlap data as a new dataframe

#Divide each column of overlapData by respective totalLines entry. Multiply by 100 and and save the percentage in a new dataframe

for(i in 1:length(overlapData$totalCpG)) {
  proportionData$totalCpG[i] <- (overlapData$totalCpG[i]/overlapData$totalCpG[6])*100
}

for(i in 1:length(overlapData$methylatedCpG)) {
  proportionData$methylatedCpG[i] <- (overlapData$methylatedCpG[i]/overlapData$methylatedCpG[6])*100
}

for(i in 1:length(overlapData$DML)) {
  proportionData$DML[i] <- (overlapData$DML[i]/overlapData$DML[6])*100
}
```

```{r}
proportionData <- proportionData[-6,] #Remove totalLines row
head(proportionData) #Confirm changes
```

## Total CpGs vs. methylated CpGs

```{r}
totalVersusMethylatedProportions <- proportionData[,1:2] #Subset data
head(totalVersusMethylatedProportions) #Confirm subset
```

```{r}
#pdf("2019-04-10-Total-Versus-Methylated-CpGs.pdf", height = 8.5, width = 11)
barplot(t(totalVersusMethylatedProportions), 
        beside = TRUE, 
        axes = FALSE, 
        names.arg = c("Exons", "Introns", "TE", "Promoters", "None")) #Create a grouped barplot (beside = TRUE) using a transposed version of the proportion data. Use axes = FALSE to remove the y-axis and names.arg to set labels on the x-axis.
axis(side = 2, at = seq(0, 45, by = 15), las = 2, col = "grey80")
mtext(side = 2, "Proportion CpGs", line = 2.5) #Add y-axis label
legend("topright", 
       legend = c("All", "Methylated"),
       pch = 16,
       col = c("grey20", "grey80"),
       bty = "n") #Place a legend in the top right of the figure with no box
#dev.off()
```

## Methylated CpGs vs. DML

```{r}
methylatedVersusDMLProportions <- proportionData[,2:3] #Subset data
head(methylatedVersusDMLProportions) #Confirm subset
```

```{r}
#pdf("2019-04-10-Methylated-CpGs-Versus-DML.pdf", height = 8.5, width = 11)
barplot(t(methylatedVersusDMLProportions), 
        beside = TRUE, 
        axes = FALSE, 
        names.arg = c("Exons", "Introns", "TE", "Promoters", "None")) #Create a grouped barplot (beside = TRUE) using a transposed version of the proportion data. Use axes = FALSE to remove the y-axis and names.arg to set labels on the x-axis.
axis(side = 2, at = seq(0, 60, by = 15), las = 2, col = "grey80")
mtext(side = 2, "Proportion Loci", line = 2.5)
legend("topright", 
       legend = c("Methylated CpGs", "DML"),
       pch = 16,
       col = c("grey20", "grey80"),
       bty = "n") #Place a legend in the top right of the figure with no box
#dev.off()
```