---
title: "Differentially Methylated Gene Analysis"
author: "Yaamini Venkataraman"
date: "8/19/2019"
output: github_document
---

In this script, I will identify genes that are differentially methylated in response to *C. virginica* exposure to experimental ocean acidification conditions. I am roughly replicating an analysis from [this paper](https://advances.sciencemag.org/content/4/6/eaar8028) and [this preprint](https://www.biorxiv.org/content/early/2018/02/21/269076). .I will annotate coverage output from `bismark` with gene information, calculate median methylation for each gene, then perform t-tests for each gene. For differentially methylated genes (DMG), I will conduct a functional enrichment with GO-MWU.

# Set up R Markdown document

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Install packages

```{r}
#install.packages("dplyr")
require(dplyr)
```

# Obtain session information

```{r}
sessionInfo()
```

# Create annotated gene list

## Add gene track

```{r}
geneList <- read.delim("../2019-05-13-Generating-Genome-Feature-Tracks/C_virginica-3.0_Gnomon_gene_sorted_yrv.gff3", header = FALSE) #Import gene gff3
colnames(geneList) <- c("gene-chr", "Gnomon", "gene", "gene-start", "gene-end", "V6", "V7", "V8", "V9") #Rename columns
head(geneList) #Confirm import
```

```{r}
geneList <- geneList[,-c(6:8)] #Remove columns 6-8
head(geneList) #Confirm changes
```

### Import gene ID

```{r}
geneDescriptions <- read.delim2("../2018-12-02-Gene-Enrichment-Analysis/2019-06-20-Gene-Descriptions.txt", header = FALSE, col.names = c("gene-number", "gene-ID", "V3", "V4", "V5", "V6", "V7", "V8", "V9")) #Import document with gene descrptions. Name 9 columns so all rows are imported with 9 columns
head(geneDescriptions) #Confirm import
```

```{r}
geneID <- geneDescriptions$gene.ID #Isolate gene ID column
geneID <- gsub("Dbxref=GeneID:", "", geneID) #Remove extraneous characters
head(geneID) #Confirm changes
```

### Merge gene location and gene ID

```{r}
geneList <- data.frame(geneList, geneID) #Replace geneList dataframe with geneList and geneID
geneList <- geneList[,-6] #Remove extra column
colnames(geneList) <- c("gene-chr", "Gnomon", "gene", "gene-start", "gene-end", "gene-ID") #Rename columns to match previous naming conventions
head(geneList) #Confirm merge
```

## Add mRNA track

```{r}
mRNAList <- read.delim("../2019-05-13-Generating-Genome-Feature-Tracks/C_virginica-3.0_Gnomon_mRNA_yrv.gff3", header = FALSE) #Import mRNA track
mRNAList <- mRNAList[,-c(6,8)] #Remove extra columns
colnames(mRNAList) <- c("mRNA-chr", "Gnomon", "mRNA", "mRNA-start", "mRNA-end", "mRNA-strand", "V9") #Rename columns
head(mRNAList) #Confirm import
```

### Import mRNA descriptions

```{r}
mRNADescriptions <- read.delim2("../2018-12-02-Gene-Enrichment-Analysis/2019-06-20-mRNA-Descriptions.txt", header = FALSE, col.names = paste("V", 1:78, sep = "")) #Import document with gene descrptions. Name 78 columns so all rows are imported with 78 columns
head(mRNADescriptions) #Confirm import
```

```{r}
geneIDmRNAList <- mRNADescriptions$V3 #Isolate gene ID column
geneIDmRNAList <- gsub("Dbxref=GeneID:", "", geneIDmRNAList) #Remove extraneous characters
head(geneIDmRNAList) #Confirm changes
```

```{r}
genbankIDmRNAList <- mRNADescriptions$V4 #Isolate Genbank ID
genbankIDmRNAList <- gsub("Genbank:", "", genbankIDmRNAList) #Remove extraneous characters
head(genbankIDmRNAList) #Confirm changes
```

### Import isolated mRNA products

```{r}
mRNAProducts <- read.delim2("../2018-12-02-Gene-Enrichment-Analysis/2019-06-20-mRNA-Products.txt", header = FALSE, col.names = paste("V", 1:16, sep = "")) #Import document with gene descrptions. Name 16 columns so all rows are imported with 16 columns.
head(mRNAProducts) #Confirm import
```

```{r}
mRNAProducts <- mRNAProducts$V1 #Extract the first column with product information
```

### Reformat mRNA List

```{r}
mRNAList <- data.frame(mRNAList, geneIDmRNAList, genbankIDmRNAList, mRNAProducts) #Replace mRNAList dataframe with mRNAList, geneIDmRNAList, genbankIDmRNAList, and mRNAProducts
mRNAList <- mRNAList[,-c(6:7)] #Remove extra columns
colnames(mRNAList) <- c("mRNA-chr", "Gnomon", "mRNA", "mRNA-start", "mRNA-end", "gene-ID", "Genbank", "Product") #Rename columns to match previous naming conventions
head(mRNAList) #Confirm merge
```

## Merge gene and mRNA tracks

```{r}
geneAnnotations <- unique(merge(x = geneList, y = mRNAList, by = "gene-ID")) #Merge gene and mRNA lists by gene-ID column. I will only have genes with mRNA in the final dataframe. Additionally, keep only unique entries (remove any merging artifacts)
length(geneAnnotations$`gene-ID`) #There are 60,201 mRNA, so this is a good sign!
```

```{r}
geneAnnotations <- geneAnnotations[,c(2, 5:6, 10:11, 1, 12, 13)] #Reorganize columns
head(geneAnnotations) #Confirm reorganization
```

## Retain the longest mRNA for each gene

When looking at differentially methylated genes, Liew et al. used the longest mRNA for each gene. I will do the same thing.

```{r}
geneAnnotations$mRNALength <- as.numeric(geneAnnotations$`mRNA-end` - geneAnnotations$`mRNA-start`) #Calculate the length of each mRNA and save in a new column
head(geneAnnotations) #Confirm length calculation
```

```{r}
attach(geneAnnotations) #Attach dataframe
geneAnnotations <- geneAnnotations[order(`gene-chr`, `gene-start`, -mRNALength),] #Sort by gene chromosome, then gene start position, and finally mRNA length. The minus sign indicates the dataframe should be sorted in descending order (longer mRNA length first)
detach(geneAnnotations) #Detach dataframe
head(geneAnnotations) #Confirm sorting
```

```{r}
write.table(geneAnnotations, "2019-08-14-All-Annotated-Gene-and-mRNA.txt", quote = FALSE, row.names = FALSE) #Save ordered file
```

```{bash}
#For the 6th field [$6] (gene ID), remove all duplicate entries in a csv file (-F","). These entries are saved and can be extracted as a different output if needed. 
#Save output to a new file
awk -F"" '!x[$6]++' 2019-08-14-All-Annotated-Gene-and-mRNA.txt \
> 2019-08-14-All-Annotated-Gene-and-mRNA-Unique.csv
```

```{bash}
head 2019-08-14-All-Annotated-Gene-and-mRNA-Unique.csv #Check file
```

```{r}
geneAnnotationsCondensed <- read.csv("2019-08-14-All-Annotated-Gene-and-mRNA-Unique.csv", header = TRUE) #Import condensed information
length(geneAnnotationsCondensed$gene.chr) #Number of mRNA from unique genes.
```

## Add Uniprot and GOterm information

```{r}
blastResults <- read.delim("../2018-12-02-Gene-Enrichment-Analysis/2018-09-11-Transcript-Uniprot-blastx-codeIsolated.txt", header = FALSE, sep = "\t", dec = ".") #Import blast results
blastResults <- blastResults[,c(1, 3)] #Keep only relevant columns
colnames(blastResults) <- c("Genbank", "Uniprot") #Rename columns
head(blastResults) #Confirm column headers
```

```{r}
geneAnnotationsCondensedUniprot <- unique(merge(x = geneAnnotationsCondensed, y = blastResults, by = "Genbank", all.x = TRUE, sort = FALSE)) #Add Uniprot Accession code to gene annotations. Keep all rows even if they don't have matching Uniprot codes. Do not resort the output. Keep only unique entries (remove merging artifacts)
head(geneAnnotationsCondensedUniprot) #Confirm merge
```

```{r}
uniprotGOTerms <- read.delim("../2018-12-02-Gene-Enrichment-Analysis/uniprot-reviewed_yes.tab", header = TRUE) #Import tab-delimited file with header
uniprotGOTerms <- uniprotGOTerms[,c(1,8)] #Retain Entry and GOterm columns
colnames(uniprotGOTerms) <- c("Uniprot", "GO") #Rename columns
head(uniprotGOTerms) #Confirm import
```

```{r}
geneAnnotationsCondensedUniprotGOterms <- unique(merge(x = geneAnnotationsCondensedUniprot, y = uniprotGOTerms, by = "Uniprot", all.x = TRUE)) #Add GOterms to gene annotations. Keep all rows even if they don't have matching Uniprot codes. Keep only unique entries (remove merging artifacts)
head(geneAnnotationsCondensedUniprotGOterms, n = 2) #Confirm merge. The output is sorted by Uniprot codes, so I'll need to resort. It would have been sorted by Uniprot code order even if I had used sort = FALSE.
```

```{r}
geneAnnotationsCondensedUniprotGOterms <- geneAnnotationsCondensedUniprotGOterms[,c(3:8, 2, 10, 9, 1, 11)] #Reorganize columns
attach(geneAnnotationsCondensedUniprotGOterms) #Attach dataframe
geneAnnotationsCondensedUniprotGOterms <- geneAnnotationsCondensedUniprotGOterms[order(gene.chr, gene.start),] #Sort by gene chromosome, then start position
detach(geneAnnotationsCondensedUniprotGOterms) #Detatch dataframe
head(geneAnnotationsCondensedUniprotGOterms, n = 5) #Confirm reorganization and sorting
```

I will now save the gene information as a new file. This information will be used to annotate coverage files.

```{r}
write.csv(geneAnnotationsCondensedUniprotGOterms, "2019-08-14-Annotated-Gene-Longest-mRNA-Uniprot-GOterms.csv", quote = FALSE, row.names = FALSE) #Save file
```

```{r}
write.table(geneAnnotationsCondensedUniprotGOterms, "2019-08-14-Annotated-Gene-Longest-mRNA-Uniprot-GOterms.txt", sep = "\t", col.names = FALSE, row.names = FALSE, quote = FALSE) #Also saving a tab-delimited version
```

# Annotate coverage files

## Download 5x coverage files

Liew et al. used raw coverage files from `bismark` for their analysis, but I will use manipulated coverage files. In [this Jupyter notebook](https://github.com/fish546-2018/yaamini-virginica/blob/master/notebooks/2019-03-07-Generating-Coverage-Tracks.ipynb), I filtered loci with 5x coverage and saved that information in bedGraphs. These bedGraphs will be used in the DMG analysis.

```{bash, echo = FALSE}
wget -r -l1 --no-parent -A_5x.bedgraph https://gannet.fish.washington.edu/spartina/2018-10-10-project-virginica-oa-Large-Files/2019-03-07-Yaamini-Virginica-Repository/analyses/2019-03-07-IGV-Verification/ #Download 5x bedgraphs from gannet
```

```{bash}
mv gannet.fish.washington.edu/spartina/2018-10-10-project-virginica-oa-Large-Files/2019-03-07-Yaamini-Virginica-Repository/analyses/2019-03-07-IGV-Verification/* . #Move files to present working directory. Where they are stored mimics file structure on gannet
```

```{bash}
rm -r gannet.fish.washington.edu #Remove gannet directory
```

```{bash}
head zr2096_1_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph #Columns: chr, start, end, percent methylation. Tab-delimited.
```

## Create new files without percent methylation information

To match loci with the gene they're in, I'm going to use `intersectBed` from `bedtools`. Before I can use files with `intersectBed`, I need to remove the last column from the coverage files, percent methylation. This is because `intersectBed` will only accept chromosome and positions, and will not accept columns with numbers that are not positions.

```{bash}
#For each file that ends in bedgraph
#Print the first three columns, tab-delimited, and save the columns in a new file that has the same base name and ends in posOnly
for f in *bedgraph
do
  awk '{print $1"\t"$2"\t"$3}' ${f} > ${f}.posOnly
done
```

```{bash}
find *posOnly #Confirm files were created
```

```{bash}
head zr2096_5_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.posOnly #Confirm file formatting
```

## Run `intersectBed`

```{bash}
#For each file that ends in posOnly
#Use intersectBed to find where loci and genes intersect, allowing loci to be mapped to annotated genes
#wb: Print all lines in the second file
#a: file that ends in posOnly
#b: annotated gene list
#Save output in a new file that has the same base name and ends in -Annotated.txt
for f in *posOnly
do
  /Users/Shared/bioinformatics/bedtools2/bin/intersectBed \
  -wb \
  -a ${f} \
  -b 2019-08-14-Annotated-Gene-Longest-mRNA-Uniprot-GOterms.txt \
  > ${f}-Annotated.txt
done
```

```{bash}
find *Annotated.txt
```

```{bash}
head zr2096_1_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.posOnly-Annotated.txt #Confirm file is annotated
```

## Add back percent methylation information

After annotating coverage files with `intersectBed`, I need to add percent methylation information back to the annotated files. I need the percent methylation data to calculate median methylation by gene. I could import 20 files into R, or I could use a series of `bash` commands.

```{bash}
#For each file that ends in bedgraph (the original coverage files)
#Print the first three columns (chr, start, end) with dashes in between, then the rest of the columns in the file
#Save the file with the base name + .sorted
for f in *bedgraph
do
  awk '{print $1"-"$2"-"$3"\t"$0}' ${f} \
  | sort -k1,1 \
  > ${f}.sorted
done
```

```{bash}
find *bedgraph.sorted
```

```{bash}
head zr2096_1_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.sorted
```

```{bash}
#For each file that ends in Annotated.txt
#Print the first three columns (chr, start, end) with dashes in between, then the rest of the columns in the file
#Save the file with the base name + .sorted
for f in *Annotated.txt
do
  awk '{print $1"-"$2"-"$3"\t"$0}' ${f} \
  | sort -k1,1 \
  > ${f}.sorted
done
```

```{bash}
find *Annotated.txt.sorted
```

```{bash}
head zr2096_1_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.posOnly-Annotated.txt.sorted
```

```{bash}
for n in {1..10}
do
  join -j 1 -t $'\t' \
  "zr2096_$n_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.sorted" \
  "zr2096_$n_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.posOnly-Annotated.txt.sorted" \
  > "sample_$n_annotated-percentMeth"
done
```

```{bash}
for f in *bedgraph
do
  join -j 1 -t $'\t' \
  ${f}.sorted \
  ${f}.posOnly-Annotated.txt.sorted \
  > ${f}.annotated-percentMeth
done
```

```{bash}
find *annotated-percentMeth
```

```{bash}
head zr2096_1_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.bedgraph.annotated-percentMeth
```








